pkg/ingester/ingester.go

```diff
@@ -1120,6 +1146,28 @@ func (i *Ingester) Push(ctx context.Context, req *cortexpb.WriteRequest) (*corte
 	// process it before samples. Otherwise, we risk returning an error before ingestion.
 	ingestedMetadata := i.pushMetadata(ctx, userID, req.GetMetadata())
 
+	type labelsCounter struct {
+		reasonCounter map[string]int
+		lbls          labels.Labels
+	}
+
+	labelSetCounters := make(map[uint64]*labelsCounter)
+
+	updateLabelSetCounter := func(matchedLabelSetLimits []validation.LimitsPerLabelSet, reason string) {
+		for _, l := range matchedLabelSetLimits {
+			if c, exists := labelSetCounters[l.Hash]; exists {
+				c.reasonCounter[reason]++
+			} else {
+				labelSetCounters[l.Hash] = &labelsCounter{
+					reasonCounter: map[string]int{
+						reason: 1,
+					},
+					lbls: l.LabelSet,
+				}
+			}
+		}
+	}
+
```

Comment:
Could this be a function of the `(l *Limits)` struct?

Reply:
Make it a function of the struct